### PR TITLE
178: Allow psd2 info to be checked when collecting userName

### PR DESCRIPTION
**Pull request by BohoCode**
Fixes https://github.com/openbanking4-dev/spring-security-multi-auth/issues/178

### CODE REVIEW

File: starter/src/main/java/dev/openbanking4/spring/security/multiauth/configurers/collectors/PSD2Collector.java [hunk 1]
```diff
@@ -42,8 +42,19 @@
 public class PSD2Collector extends X509Collector {
 
     @Builder(builderMethodName = "psd2Builder")
-    public PSD2Collector(String collectorName, UsernameCollector usernameCollector, AuthoritiesCollector authoritiesCollector, CertificateHeaderFormat collectFromHeader, String headerName) {
-        super(collectorName, usernameCollector, certificatesChain -> {
+    public PSD2Collector(String collectorName, PSD2UsernameCollector usernameCollector,
+                         AuthoritiesCollector authoritiesCollector, CertificateHeaderFormat collectFromHeader,
+                         String headerName) {
+        super(collectorName, certificatesChain -> {
```
Comment (matt-wills) [inline starter/src/main/java/dev/openbanking4/spring/security/multiauth/configurers/collectors/PSD2Collector.java]: The lambdas here make it difficult to see where the call to the super constructor starts and ends, and which parameters are being provided.

Instead we could extract them into methods and have:

```
    super(collectorName,
          getUsernameCollector(usernameCollector),
          getAuthoritiesCollector(authoritiesCollector),
          collectFromHeader,
          headerName);
```

IntelliJ will give you the option to extract them into methods, but here they are for your convenience:

```
    private static UsernameCollector getUsernameCollector(PSD2UsernameCollector usernameCollector) {
        return certificatesChain -> {
            try {
                Psd2CertInfo psd2CertInfo = new Psd2CertInfo(certificatesChain);
                return usernameCollector.getUserName(certificatesChain, psd2CertInfo);
            } catch (InvalidPsd2EidasCertificate | InvalidEidasCertType invalidPsd2EidasCertificate) {
                log.warn("Certificate found couldn't be parsed as a PSD2 certificate. Certificate will be ignored",
                        invalidPsd2EidasCertificate);
                return null;
            }
        };
    }

    private static X509Collector.AuthoritiesCollector getAuthoritiesCollector(AuthoritiesCollector authoritiesCollector) {
        return certificatesChain -> {
            Set<GrantedAuthority> authorities = new HashSet<>();
            try {
                Psd2CertInfo psd2CertInfo = new Psd2CertInfo(certificatesChain);
                log.trace("Verify if certificate is a QWAC");
                if (psd2CertInfo.isPsd2Cert()
                        && psd2CertInfo.getEidasCertType().isPresent()
                        && psd2CertInfo.getEidasCertType().get().equals(EidasCertType.WEB)) {

                    //Map PSD2 roles
                    Optional<Psd2QcStatement> psd2QcStatementOpt = psd2CertInfo.getPsd2QCStatement();
                    if (psd2QcStatementOpt.isPresent()) {
                        Psd2QcStatement psd2QcStatement = psd2QcStatementOpt.get();
                        log.trace("Founds PSD2 QC Statement {}", psd2QcStatement);
                        authorities.addAll(authoritiesCollector.getAuthorities(certificatesChain, psd2CertInfo, psd2QcStatement.getRoles()));
                    } else {
                        log.trace("No PSD2 QC Statement found");
                        authorities.addAll(authoritiesCollector.getAuthorities(certificatesChain, psd2CertInfo, null));
                    }
                } else {
                    if (log.isTraceEnabled()) {
                        if (!psd2CertInfo.isPsd2Cert()) {
                            log.trace("Not a PSD2 cert");
                        } else if (psd2CertInfo.getEidasCertType().isEmpty()) {
                            log.trace("Is a PSD2 certs but no EIDAS cert type");
                        } else if (psd2CertInfo.getEidasCertType().isEmpty()) {
                            log.trace("Is a PSD2 certs with EIDAS cert type {} but it's not a QWAC", psd2CertInfo.getEidasCertType().get());
                        }
                    }
                    authorities.addAll(authoritiesCollector.getAuthorities(certificatesChain, null, null));
                }
            } catch (InvalidPsd2EidasCertificate | InvalidEidasCertType invalidPsd2EidasCertificate) {
                log.warn("Certificate founds couldn't be parsed as a PSD2 certificate. Will ignore the certificate", invalidPsd2EidasCertificate);
            }

            return authorities;
        };
    }
```

File: starter/src/test/java/dev/openbanking4/spring/security/multiauth/configurers/collectors/PSD2CollectorTest.java [hunk 2]
```diff
@@ -127,9 +133,33 @@
...
+        // Build a collector that will not return a username if the cert is not an eidas WEB cert.
         this.psd2Collector = PSD2Collector.psd2Builder()
                 .collectorName("psd2-for-test")
-                .usernameCollector(certificatesChain -> certificatesChain[0].getSubjectDN().getName())
+                .usernameCollector((certificatesChain, psd2CertInfo) -> {
+                    String userName = null;
+                    if(psd2CertInfo != null && psd2CertInfo.isPsd2Cert()){
+                        Optional<EidasCertType> eidasCertType = psd2CertInfo.getEidasCertType();
+                        if (eidasCertType.isPresent()) {
+                            if( eidasCertType.get() == EidasCertType.WEB ) {
```
Comment (matt-wills) [inline starter/src/test/java/dev/openbanking4/spring/security/multiauth/configurers/collectors/PSD2CollectorTest.java]: Minor - the formatting of the if statements is a bit inconsistent in each of these classes. The formatter dictates a space after the `if`.

File: starter/src/test/java/dev/openbanking4/spring/security/multiauth/configurers/collectors/PSD2CollectorTest.java [hunk 3]
```diff
@@ -127,9 +133,33 @@
...
+                                if (certificatesChain[0] != null) {
+                                    userName = certificatesChain[0].getSubjectDN().getName();
+                                } else {
+                                    throw new IllegalArgumentException("certificateChain must not contain null " +
+                                            "entries");
+                                }
+                            } else {
+                                log.debug("Expected eidasCertType to be WEB. It was '{}'. Will not collect username",
```
Comment (matt-wills) [inline starter/src/test/java/dev/openbanking4/spring/security/multiauth/configurers/collectors/PSD2CollectorTest.java:150]: There are quite a few indented if statements, which makes it a little hard to read (and determine which else belong to which if). I guess you find the log statements helpful if there are problems, but of course most of the time they will be ignored.

I wonder if it's worth combining them and failing early like this?

```
if (eidasCertType.isEmpty() || eidasCertType.get() != EidasCertType.WEB || certificatesChain[0] == null) {
   fail("Could not extract username from certificate chain");
} else {
   userName = certificatesChain[0].getSubjectDN().getName();
}
```

**Review by matt-wills — DISMISSED** [review]

File: starter/src/main/java/dev/openbanking4/spring/security/multiauth/configurers/collectors/PSD2Collector.java (hunk 1 above)
Comment (BohoCode) [reply starter/src/main/java/dev/openbanking4/spring/security/multiauth/configurers/collectors/PSD2Collector.java]: Thanks Matt

File: starter/src/test/java/dev/openbanking4/spring/security/multiauth/configurers/collectors/PSD2CollectorTest.java (hunk 3 above)
Comment (BohoCode) [reply starter/src/test/java/dev/openbanking4/spring/security/multiauth/configurers/collectors/PSD2CollectorTest.java:150]: Fixed (ish)

File: starter/src/main/java/dev/openbanking4/spring/security/multiauth/configurers/collectors/PSD2Collector.java (hunk 1 above)
Comment (BohoCode) [reply starter/src/main/java/dev/openbanking4/spring/security/multiauth/configurers/collectors/PSD2Collector.java]: Fixed

File: starter/src/test/java/dev/openbanking4/spring/security/multiauth/configurers/collectors/PSD2CollectorTest.java (hunk 3 above)
Comment (matt-wills) [reply starter/src/test/java/dev/openbanking4/spring/security/multiauth/configurers/collectors/PSD2CollectorTest.java:150]: I'll take your word for it, but I'm not sure what you've changed! :-)

**Review by matt-wills — APPROVED** [review]

**Pull request merged**